### DISCUSSION TITLE
Yuki Theme Features

### DISCUSSION THREAD

**Discussion by Dragon-0609** [Dragon-0609/Yuki-Theme#1] [Show and tell]
## 👋 Welcome! Here I'll show and tell what I have achieved.
You can tell me your opinion.

---

Now, the plugin can show background image in all tabs. I have found way to track active tab. You can see it here:

![screenshoot](https://user-images.githubusercontent.com/80271099/142404840-1ac0d5a0-d2a3-417b-abee-095cf75cd8dc.png)

![screenshot 2](https://user-images.githubusercontent.com/80271099/142404548-f2749951-ab8d-4621-b81a-ba12bead97c8.png)

I have integrated `Yuki Theme` as plugin, so I'll think about refreshing theme of `PascalABC.NET` on seleceting theme at runtime. At the moment, the program reads colors on opening.

**Comment by Dragon-0609** [comment]
> I'll assume you have some way to detect when a theme is changed, to set the appropriate background, right or is that something the user will have to manually apply?

I can detect theme changing. Actually, theme changes in `Yuki Theme`, not in Pascal. So, it's easy to detect theme changing. I have problems with background images. Although I can set background image, I must choose way to select background image. <br>
I think about integrating all `Doki Theme` color schemes, so background images I can hardcode. Or I can write in somewhere of theme file 
the path to the bg image. <br>
Also, background images are shown because of my plugin. I think, I can let user to select custom bg image for custom themes. I can't let them to change default themes.

**Reply by Unthrottled** [reply]
> I can detect theme changing. Actually, theme changes in Yuki Theme, not in Pascal. So, it's easy to detect theme changing.

That's good, I like the user experience of when a theme changes, that the assets change automatically with the theme change.

> I think about integrating all Doki Theme color schemes, so background images I can hardcode. Or I can write in somewhere of theme file
the path to the bg image. 

From what I understand right now, `Yuki Theme` allows users to import/create various color schemes  from the `icls` files exported from a JetBrains IDE, right? Are you planning on integrating some default themes (including themes in the Doki Theme suite)?

**Reply by Dragon-0609** [reply]
> From what I understand right now, Yuki Theme allows users to import/create various color schemes from the icls files exported from a JetBrains IDE, right? 

Yes. I have put a parser for JetBrains IDE color scheme (icls).

> Are you planning on integrating some default themes (including themes in the Doki Theme suite)?

I have integrated `Darcula` (default color scheme of JetBrains IDE). I can integrate themes of Doki Theme. It is a piece of cake. Now, default schemes are in `default` category. I'll add `Doki Theme` to its own category.

**Reply by Dragon-0609** [reply]
It'll be like here:
![screenshot](https://user-images.githubusercontent.com/80271099/142741722-7388411b-77ef-490b-9e24-f01ea3bac901.png)

**Reply by Unthrottled** [reply]
> I have integrated Darcula (default color scheme of JetBrains IDE). I can integrate themes of Doki Theme. It is a piece of cake. Now, default schemes are in default category. I'll add Doki Theme to its own category.

Got it, thanks for documenting what this tool does. What I understand is that it provides users a way to export configurations directly to the PascalABC.net (That's the IDE, I think), that way when the user starts the program again it is themed.
 
So if I wanted wanted say, Zero Two Dark, I could import the icls, choose my background image, set the position. Then the it would be themed like Zero Two Dark?

**Reply by Dragon-0609** [reply]
> What I understand is that it provides users a way to export configurations directly to the PascalABC.net (That's the IDE, I think), that way when the user starts the program again it is themed.

Yes. If you use the program version (current version, it's .exe), you have to export theme. After that, program is themed, but, just it's text editor is themed. Other things won't be themed. Even Background Image won't be shown. The program version has limitations. But the plugin version (currently I'm developing, it's .dll), you just select theme and press Save. After that it'll refresh layout, so new theme can be applied. The plugin version will change color of some UI parts as Menu, StatusBar and etc. Also, plugin version will show background image.

> So if I wanted wanted say, Zero Two Dark, I could import the icls, choose my background image, set the position. Then the it would be themed like Zero Two Dark?

Yes, you can do that. After importing, the program saves it for later usages. For example, you import let's say `Zero Two Dark`, set its background image and position. Then you save it and export it. Later, you change theme to another let's say `Megumin`, but `Zero Two Dark` will be there, so you can apply it later. I mean, you import it at once, change its background image and positions, save.

**Reply by Unthrottled** [reply]
Got it, sounds like users can import themes that they want if it doesn't already exist. 

If you ever plan on integrating the Doki Theme suite, I'll be able to lend a hand with that. I've got some tooling available that makes it manageable to maintain 67 themes (Going to be 70 by December 🎄 )

Thanks for sharing!

**Reply by Dragon-0609** [reply]
> If you ever plan on integrating the Doki Theme suite, I'll be able to lend a hand with that. I've got some tooling available that makes it manageable to maintain 67 themes

Thanks. I'll notify you when I start integrating the Doki Theme suite. 

I wanna make online theme repo, so everyone can download any theme inside the program. I think, I'll add this feature in next version (Maybe in v3).

**Reply by Dragon-0609** [reply]
Before when I start integrating the Doki Theme suite, I'll make a feature to auto-import themes of JetBrains IDE (icls) and their background image.

For example. here's the directory:
```
Themes
|──  Echidna.icls 	──|_____ 
|──  Echidna.png	──|	|── Echidna.yukitheme
|──  Ishtar_Dark.icls	──|_____ 
|──  Ishtar_Dark.png	──|     |── Ishtar_Dark.yukitheme
|──  Ram.icls		──|_____ 
|──  Ram.png		──|     |── Ram.yukitheme
|──  Rem.icls		──|_____ 
|──  Rem.png		──|     |── Rem.yukitheme
|──  .........


```

It'll import themes automatically. You just select the directory and that is all.

**Reply by Unthrottled** [reply]
> I'll make a feature to auto-import themes of JetBrains IDE (icls) and their background image.

Cool that sounds neat! 

Unfortunately that isn't going to work for the Doki Theme very well. It will work for some themes, but there isn't going to be an easy way to link background images and `icls` files by name alone.

I could put something together that generates [.yukitheme](https://raw.githubusercontent.com/Dragon-0609/Yuki-Theme/main/Yuki%20Theme/Themes/Github%20Dark.yukitheme) (I assume this is the expected output that is built from an ICLS file), and skips the need for parsing a ICLS file.
Just as long as a base theme is defined, the rest is fairly straight forward. 

Anyways, continue doing what you are going to do, if you are going to take a small part of the doki-theme go for it. But if you want it all, some things will need to be done (once I get the time :)

**Reply by Dragon-0609** [reply]
> I could put something together that generates .yukitheme (I assume this is the expected output that is built from an ICLS file), and skips the need for parsing a ICLS file.

If you wanna do that, in [`JetBrainsParser`](https://github.com/Dragon-0609/Yuki-Theme/blob/main/Yuki%20Theme/Parsers/JetBrainsParser.cs) you can see from where the program gets colors. The name of function is `getName`, it is in line `238`. Let me know, when you want to do that and I'll push new version to the repo, so you will be able to add background images. Also, the program saves background image and XML to the `*.yukitheme` file. That's why you'll need new version.

**Reply by Unthrottled** [reply]
Hey! So I am back. I had a few questions before I decide whether on what I want to do next. 

For starters, kudos on all of the stuff you have done, very impressive! Secondly, I am fine with this plugin adopting/bundling the Doki Theme suite. It raises a few questions, but I'll see if I need to ask them. 

If you are going to use my themes, I am willing to entertain with the idea of collaborating on a `doki-theme-pascalABC`,  I'll help out with maintenance if that is the case. 

If you wanted to run your own show with the Yuki Theme, that's cool as well. In that case I'll leave you be, and leave the maintenance in your hands.

So what is your long term vision for this application?

**Reply by Dragon-0609** [reply]
Hi. 
> I am willing to entertain with the idea of collaborating on a doki-theme-pascalABC

I'll be thankful if you help me. If I have understood right, you want to rename `Yuki Theme` to `doki-theme-pascalABC`, haven't I? Well, I'm not against of it. It will be cool!

> If you wanted to run your own show with the Yuki Theme

I wanted to make copy of The Doki Theme for pascal. If it will be a part of The Doki Theme, I'll be happy.

> I'll help out with maintenance if that is the case.

At the moment I'm working over v4, which will include CLI version and some integration with UI. In the v5, I want to change the most of icons from UI to icons of JetBrains.
It will also include `Dynamically changing icons of the IDE`. For example, If I change theme, icons will change their colors, like in IntelliJ IDEA.
I need your help for that version.

> So what is your long term vision for this application?

At the moment, it isn't popular. But by the time, more people who use Pascal will know it. I don't know, how many people will use it, but, at the first position, I have made this plugin for myself. So, I'll try to make IDE of Pascal better with this app.

**Reply by Unthrottled** [reply]
Thank you for getting back to me on those questions.

> If I have understood right, you want to rename Yuki Theme to doki-theme-pascalABC, haven't I? 

You don't have to do anything with the repository yet.

I will eventually take some time to understand what you have set up so far. 

My first question is, how are you currently maintaining the Doki Themes, is it a manual process or automated process? You do not need to give me the details of how it is done, I just want to know if you have been manually updating things for each theme or if there is an automated script that builds the assets.

**Reply by Dragon-0609** [reply]
> how are you currently maintaining the Doki Themes, is it a manual process or automated process?

Actually, it is mixed. In 2nd version, I added Doki Theme Parser to `auto import from directory`. But when I worked over 3rd version, I had to refactor most of parts of theme files to show the names with ':' character. So I updated Doki Themes manually. Mostly because I was too lazy to refactor the Doki Theme Parser. 
I can make it automated by refactoring Doki Theme Parser. Not fully automated, you just select directory with themes, and after that, it saves to Themes directory, finally you have to embed them to the project to make them default.

> if there is an automated script that builds the assets.

There is no automated script for that. I have `auto-import from directory`. It isn't automated, but at least, it can save time.

**Reply by Dragon-0609** [reply]
> when I worked over 3rd version, I had to refactor most of parts of theme files to show the names with ':' character. So I updated Doki Themes manually. Mostly because I was too lazy to refactor the Doki Theme Parser.

Doki Theme Parser adds background images and stickers to themes automatically, but it can't set align and opacity of background image.  Because, it doesn't know, where it can get align and opacity. Because of that I updated Doki Themes manually.

**Reply by Dragon-0609** [reply]
Hi. I have an idea that I want to add to opportunities. I want to let the user choose and download new added Themes of `Doki Theme`. The user need not update the whole app, instead of that he downloads new themes without any problem.

Process of downloading:
 A list with new themes will be shown and the user will be able to press download. After downloading, Yuki Theme will convert new theme using Doki Theme Parser. I think I can add this opportunity, but the problem is in parameters of Background Image. I don't know where I can get the opacity and align of wallpaper.

**Reply by Unthrottled** [reply]
That was one of the things I was wondering about.

I was trying to find where all of the images & stuff where put (I was going to guess in the compiled yukitheme). Because, I would also like it if it also followed the standardized asset update process as well.

 Then I got distracted by other stuff.

Right now there isn't a single place to get the opacity's and alignments of the wallpapers. 
Things that I have run into is that the way to define anchoring and opacity differs for each application. 
So I have been saving that metadata in the application specific definitions. I've got some documentation on most of the repositories, about this https://github.com/doki-theme/doki-theme-jetbrains/blob/master/CONTRIBUTING.md#application-specific-templates . The default is 5% for dark & 10% for light.

**Reply by Dragon-0609** [reply]
> I was trying to find where all of the images & stuff where put

I think I can somehow get background image + sticker + theme definition, If I get the list of all themes. 
I can get images from [doki theme assets](https://github.com/doki-theme/doki-theme-assets). Definitions from [doki master theme](https://github.com/doki-theme/doki-master-theme). The problem is in opacity and align. 

----

I have just compared opacity and align from [doki theme jetbrains/themes](https://github.com/doki-theme/doki-theme-jetbrains/tree/2455639bb0c51a8501168e5266337c636c4c7381/buildSrc/assets/themes) with current Yuki Theme values. They're same. When I added background image's opacity, I wanted to make it same with opacity of Doki Theme JetBrains. I found [doki theme jetbrains/themes](https://github.com/doki-theme/doki-theme-jetbrains/tree/2455639bb0c51a8501168e5266337c636c4c7381/buildSrc/assets/themes), but I wasn't sure their similarity. I didn't know that the default is 5% for dark & 10% for light, so when I compared I thought that they weren't same.

Now I can get opacity and aligns. Next problem is in the structure of folders.  For example:
In [doki-master-theme](https://github.com/doki-theme/doki-master-theme/tree/master/definitions) folders arranged like:
```
azurLane/essex/dark
dragonMaid/kanna/dark/
killLaKill/ryuko/light/
and etc...
```
I want to say that to get the definition, I must type it's root folder name (E.g.: dragonMaid). Also, I must know, has it dark or light theme. This idea will take much effort to become it real. At the moment, I'm too busy to work over this idea. I'll put it to my [future versions](https://github.com/Dragon-0609/Yuki-Theme/discussions/8).
Anyway, thanks and merry Xmas!🎅

**Reply by Unthrottled** [reply]
Most of these problems that you've mentioned, I've already had to solve. So I've got tooling for some of those things. 

The only thing that I haven't done is build a plugin that automatically pulls in the theme changes without having to update the plugin. I've got something like that for stickers n stuff, but typically I've expected users to update the plugin/app to get new themes.

Anyways, I've got some time later this week and I'll continue to root around this codebase. Hope you enjoyed your xmas and have a happy new year!

**Reply by Dragon-0609** [reply]
> a plugin that automatically pulls in the theme changes without having to update the plugin.

What if you keep new themes in `Themes` directory somewhere? For example, I keep all custom themes in `Themes` directory of my app's path. The app can read themes from memory (default) and directory (`Themes`). So, new downloaded themes I'll keep in `Themes` directory. Of course, I'll add a signature during the parse of the new downloaded theme to recognize new themes as `Doki Theme`, so they will be just read-only.

> typically I've expected users to update the plugin/app to get new themes

I want to add the opportunity to let the user choose and download new added Themes of `Doki Theme` without updating the app. The reason is on updating apps. I'm lazy for that, so instead of updating the whole app, people will be able to download new themes.

**Comment by Dragon-0609** [comment]
Now the user can choose the background image. But, it can't be saved. I'll work over that.

![screenshot](https://user-images.githubusercontent.com/80271099/142627049-912377e6-a339-40b3-a35c-902faf0343df.jpg)

**Comment by Dragon-0609** [comment]
I have done. Now, a user can change background image and save it. Also, aligns also work.
The only thing left is merging `Yuki Theme` and `Yuki Theme Plugin`. At the moment, they are two different things.

**Comment by Dragon-0609** [comment]
I have put a slider for the opacity of background image.

![screenshot](https://user-images.githubusercontent.com/80271099/142779455-329d3efe-3864-445e-aaaf-b592380cead5.png)

It works properly if you set the opacity < 20. But, in the higher values, the background image stands above the text.

![screenshot](https://user-images.githubusercontent.com/80271099/142779500-035d4437-ffe7-4527-91d0-ae4e99891252.png)

**Comment by Dragon-0609** [comment]
I have found a way to refresh the layout without restarting the program.

[Here's a link to the short video](https://drive.google.com/file/d/1x40PyKM3aaEaAUi--S3tXLbLG_q4tYmf/view?usp=sharing)

**Comment by Dragon-0609** [comment]
> The only thing left is merging Yuki Theme and Yuki Theme Plugin. At the moment, they are two different things.

I have done. Now, there are two types of `Yuki Theme`: `Program (.exe)` and `Plugin (.dll)`. As I said before, program type has limitations, whenever plugin type is more powerfull.
Almost done. If you don't mind, I wanna start integrating Doki Theme suite.

**Reply by Dragon-0609** [reply]
I'll push new version to repo in  2 days.

**Reply by Unthrottled** [reply]
Go for it, I'll swing by later this year and see if I can help out with setting up something up with maintenance (if it needs to be done.).

**Comment by Dragon-0609** [comment]
I have troubles with pushing to remote repo. I'll have to solve the problem.
BTW I have integrated Doki Theme suite. Also, I have added JSON parser for doki-theme-assets, so it wasn't too difficult.

**Comment by Dragon-0609** [comment]
Finally, I uploaded new version.
In next version, I want to add stickers. Although there're colors and background images, I must add stickers. Otherwise it won't be copy of Doki Theme.

**Comment by Dragon-0609** [comment]
I have added  name on status bar and stickers.

![screenshot](https://user-images.githubusercontent.com/80271099/144752412-4a6c78ff-9705-4b09-8d50-28e9ac9b6d02.png)

Next I want to add is a CLI version.

**Comment by Dragon-0609** [comment]
After CLI, I want to add icons Intellij IDEA, like in doki theme they will change their color.

**Comment by Dragon-0609** [comment]
At the moment I'm working over the 4th version. I have added the CLI version. Also, I have made better integration with IDE. Here are screenshots to compare:
Version 3:

![bandicam 2021-12-20 00-23-03-232](https://user-images.githubusercontent.com/80271099/146688954-091a0e47-8221-4fb7-aced-6a81a4a4eca8.png)

New version:
![bandicam 2021-12-20 00-23-28-777](https://user-images.githubusercontent.com/80271099/146688960-a89d9643-356b-46d1-958b-87c54a943bbb.png)

**Comment by Dragon-0609** [comment]
Also, I added a theme switcher to the plugin. It is shown when the user presses Ctrl + ` (below Esc). It looks like:

![bandicam 2021-12-20 11-34-30-839](https://user-images.githubusercontent.com/80271099/146725138-1c02297c-1446-4b79-9598-726a8c38c9b1.png)

**Comment by Dragon-0609** [comment]
Now PascalABC.NET looks like below. I have found a way to completely change tab colors.
![bandicam 2021-12-20 17-57-21-330](https://user-images.githubusercontent.com/80271099/146771257-51bc28d3-ec68-4703-a41c-093fa04e982e.png)

**Comment by Dragon-0609** [comment]
Today I made my resolutions to 2022. In the resolutions I marked the end of active development of `Yuki Theme` in begin of April. I'll finish all versions written [here](https://github.com/Dragon-0609/Yuki-Theme/discussions/8) till April. After that I'll switch to passive development. By that I mean I'll add anything to `Yuki Theme` rarely. In passive development updates will be released by effort of community.
NOTE. IN APRIL I WON'T CLOSE THE PROJECT. I WILL JUST RARELY WORK OVER IT. SO UPDATES WILL BE RELEASED RARELY.

**Comment by KONFEDERATE** [comment]
привет, почему то у меня не получается установит плагин, можешь помочь?

**Reply by Dragon-0609** [reply]
Привет. Можете описать, что вы делаете и почему решили, что плагин не работает?
[Тут](https://dragon-0609.github.io/Yuki-Theme/guide/installation) есть инструкция с скриншотами.

**Comment by KONFEDERATE** [comment]
я делал обсолютно так же как в инструкции просто ничего не появляется  в программе 27.10.2022, 05:16, "Dragon-LV" ***@***.***>: Привет. Можете описать, что вы делаете и почему решили, что плагин не работает?Тут есть инструкция с скриншотами.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

**Reply by Dragon-0609** [reply]
Какая версия паскаль у вас?

**Comment by Dragon-0609** [comment]
Hello everyone. You might be wondering why so long I haven't worked on Yuki Theme. The main reason is I don't have enough motivation to keep on working on the project. I rested to get some motivation but haven't collected enough yet. The second reason is at the moment I'm working on a paid project, so I can't give my full attention to this project. Although I'm working on the 2nd project, I have some free time, but I can't force myself to work on the project. Also, I'm about to finish 2nd project.
So, I hope I'll want to improve Yuki Theme after finishing the 2nd project.
